contrib/valkey-go/valkey.go

```diff
@@ -0,0 +1,369 @@
+	return command, statement, size
+}
+
+func processMultiCompleted(multi ...valkey.Completed) (command, statement string, size int) {
+	cmds := make([]commander, len(multi))
+	for i, cmd := range multi {
+		cmds[i] = &cmd
+	}
+	return processMultiCmds(cmds)
+}
+
+func processMultiCacheableTTL(multi ...valkey.CacheableTTL) (command, statement string, size int) {
+	cmds := make([]commander, len(multi))
+	for i, cmd := range multi {
+		cmds[i] = &cmd.Cmd
+	}
+	return processMultiCmds(cmds)
+}
+
+func firstError(s []valkey.ValkeyResult) error {
+	for _, result := range s {
+		if err := result.Error(); err != nil && !valkey.IsValkeyNil(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func setClientCacheTags(s tracer.Span, result valkey.ValkeyResult) {
+	s.SetTag(ext.ValkeyClientCacheHit, result.IsCacheHit())
+	s.SetTag(ext.ValkeyClientCacheTTL, result.CacheTTL())
+	s.SetTag(ext.ValkeyClientCachePTTL, result.CachePTTL())
+	s.SetTag(ext.ValkeyClientCachePXAT, result.CachePXAT())
+}
+
+type buildStartSpanOptionsInput struct {
+	command        string
+	statement      string
+	size           int
+	isWrite        bool
+	isBlock        bool
+	isMulti        bool
+	isStream       bool
+	skipRawCommand bool
+}
+
+func (c *coreClient) peerTags() []tracer.StartSpanOption {
+	ipAddr := net.ParseIP(c.host)
+	var peerHostKey string
+	if ipAddr == nil {
+		peerHostKey = ext.PeerHostname
+	} else if ipAddr.To4() != nil {
+		peerHostKey = ext.PeerHostIPV4
+	} else {
+		peerHostKey = ext.PeerHostIPV6
+	}
+	return []tracer.StartSpanOption{
+		tracer.Tag(ext.PeerService, ext.DBSystemValkey),
+		tracer.Tag(peerHostKey, c.host),
+		tracer.Tag(ext.PeerPort, c.port),
+	}
+}
+
+func (c *coreClient) buildStartSpanOptions(input buildStartSpanOptionsInput) []tracer.StartSpanOption {
+	opts := []tracer.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeValkey),
+		tracer.Tag(ext.TargetHost, c.host),
+		tracer.Tag(ext.TargetPort, c.port),
+		tracer.Tag(ext.ValkeyClientVersion, valkey.LibVer),
+		tracer.Tag(ext.ValkeyClientName, valkey.LibName),
+		tracer.Tag(ext.Component, componentName),
+		tracer.Tag(ext.SpanKind, ext.SpanKindClient),
+		tracer.Tag(ext.DBType, ext.DBSystemValkey),
+		tracer.Tag(ext.DBSystem, ext.DBSystemValkey),
+		tracer.Tag(ext.DBInstance, ext.DBSystemValkey),
+		tracer.Tag(ext.ValkeyDatabaseIndex, c.option.SelectDB),
+		tracer.Tag(ext.ValkeyClientCommandWrite, input.isWrite),
+		tracer.Tag(ext.ValkeyClientCommandBlock, input.isBlock),
+		tracer.Tag(ext.ValkeyClientCommandMulti, input.isMulti),
+		tracer.Tag(ext.ValkeyClientCommandStream, input.isStream),
+		tracer.Tag(ext.ValkeyClientCommandWithPassword, c.option.Password != ""),
+	}
+	opts = append(opts, c.peerTags()...)
+	if input.command != "" {
+		opts = append(opts, []tracer.StartSpanOption{
+			// valkeyotel tags
+			tracer.Tag("db.stmt_size", input.size),
+			tracer.Tag("db.operation", input.command),
+		}...)
+		if input.skipRawCommand {
+			opts = append(opts, tracer.ResourceName(input.command))
```

Comment:
the resource name should always be `input.statement` regardless of the value of `skipRawCommand`